File: emisor/link/hamming.py
# Capa de Enlace - Hamming SEC (Single Error Correction) | Emisor
# Responsabilidad: calcular los bits de paridad e insertarlos en las posiciones correctas.

# Funcionamiento:
#   - Los bits de paridad se ubican en posiciones que son potencias de 2: 1, 2, 4, 8, 16...
#   - Cada bit de paridad cubre las posiciones cuya representación binaria tiene un 1
#     en el bit correspondiente a esa potencia.
#   - El valor de cada bit de paridad es el XOR de todos los bits que cubre.
#   - Se debe cumplir: m + r + 1 <= 2^r
#     donde m = bits de datos, r = bits de paridad


# Calcula el número mínimo de bits de paridad r necesarios para un mensaje de m bits de datos, cumpliendo m + r + 1 <= 2^r.
def _calcular_r(m: int) -> int:
    r = 1
    while (m + r + 1) > (2 ** r):
        r += 1
    return r

# Recibe el mensaje en bits (string de '0' y '1') y retorna el frame Hamming con los bits de paridad insertados.
# Proceso:
#   1. Calcular r (bits de paridad necesarios)
#   2. Construir el frame con posiciones reservadas para paridad (valor inicial 0)
#   3. Insertar los bits de datos en las posiciones que NO son potencia de 2
#   4. Calcular el valor de cada bit de paridad mediante XOR
def codificar_hamming(bits_mensaje: str) -> str:
    m = len(bits_mensaje)
    r = _calcular_r(m)
    n = m + r # longitud total del frame

    print(f"[ENLACE-HAMMING] Bits de datos: {m}, Bits de paridad: {r}, Frame total: {n} bits")

    # Construir frame como lista; índice 0 no se usa (Hamming usa base 1)
    frame = [0] * (n + 1)

    # Insertar bits de datos en posiciones que NO son potencia de 2
    indice_dato = 0
    for pos in range(1, n + 1):
        if not _es_potencia_de_2(pos):
            frame[pos] = int(bits_mensaje[indice_dato])
            indice_dato += 1

    # Calcular el valor de cada bit de paridad
    for i in range(r):
        pos_paridad = 2 ** i
        xor_acumulado = 0
        for pos in range(1, n + 1):
            if pos & pos_paridad:
                xor_acumulado ^= frame[pos]
        frame[pos_paridad] = xor_acumulado

    # Convertir frame (sin el índice 0) a string de bits
    resultado = "".join(str(frame[i]) for i in range(1, n + 1))
    print(f"[ENLACE-HAMMING] Frame codificado: {resultado}")
    return resultado

# Retorna True si n es una potencia de 2
def _es_potencia_de_2(n: int) -> bool:
    return n > 0 and (n & (n - 1)) == 0

File: receptor/link/hamming.go
// Capa de Enlace - Hamming SEC (Single Error Correction) | Receptor
// Responsabilidad: recibir el frame Hamming, verificar integridad y corregir
// el error si existe (máximo 1 bit).
//
// Proceso de decodificación (según módulo III del curso CC3067):
//  1. Identificar las posiciones de los bits de paridad (potencias de 2)
//  2. Recalcular el XOR de cada grupo de bits cubiertos por cada paridad
//  3. Formar el "síndrome": concatenar los resultados (en orden inverso de potencia)
//  4. Si el síndrome es 0 → no hay error
//     Si el síndrome != 0 → indica la posición del bit erróneo (en base 1)
//  5. Corregir el bit en esa posición
//  6. Extraer solo los bits de datos (los que NO están en posiciones potencia de 2)

package link

import (
	"fmt"
	"math/bits"
	"strconv"
	"strings"
)

// DecodificarHamming recibe el frame Hamming como string de bits.
// Retorna los bits de datos originales (sin bits de paridad) y un error si no pudo corregir.
func DecodificarHamming(frame string) (string, error) {
	n := len(frame)
	fmt.Printf("[ENLACE-HAMMING] Frame recibido (%d bits): %s\n", n, frame)

	// Convertir el frame a slice de enteros (base 1: frame[0] no se usa)
	frameBits := make([]int, n+1)
	for i := 0; i < n; i++ {
		if frame[i] == '1' {
			frameBits[i+1] = 1
		} else {
			frameBits[i+1] = 0
		}
	}

	// Calcular cuántos bits de paridad tiene este frame
	r := calcularR(n)

	// Recalcular cada bit de paridad y construir el síndrome
	sindrome := 0
	for i := 0; i < r; i++ {
		posParidad := 1 << i // 1, 2, 4, 8...
		xorAcumulado := 0
		for pos := 1; pos <= n; pos++ {
			if pos&posParidad != 0 {
				xorAcumulado ^= frameBits[pos]
			}
		}
		// El bit i del síndrome corresponde a la paridad de posición 2^i
		if xorAcumulado != 0 {
			sindrome |= posParidad
		}
	}

	fmt.Printf("[ENLACE-HAMMING] Síndrome calculado: %d\n", sindrome)

	if sindrome != 0 {
		// El síndrome indica la posición del bit erróneo (base 1)
		posError := sindrome
		if posError > n {
			return "", fmt.Errorf("error en posición %d, fuera del rango del frame (%d bits). No se puede corregir", posError, n)
		}
		fmt.Printf("[ENLACE-HAMMING] Error detectado en posición %d. Corrigiendo...\n", posError)
		// Invertir el bit erróneo
		frameBits[posError] ^= 1
		fmt.Printf("[ENLACE-HAMMING] Bit corregido en posición %d.\n", posError)
	} else {
		fmt.Println("[ENLACE-HAMMING] No se detectaron errores.")
	}

	// Extraer solo los bits de datos (posiciones que NO son potencia de 2)
	var datosBits strings.Builder
	for pos := 1; pos <= n; pos++ {
		if !esPotenciaDe2(pos) {
			datosBits.WriteString(strconv.Itoa(frameBits[pos]))
		}
	}

	resultado := datosBits.String()
	fmt.Printf("[ENLACE-HAMMING] Bits de datos extraídos (%d bits): %s\n", len(resultado), resultado)
	return resultado, nil
}

// calcularR determina cuántos bits de paridad tiene un frame de longitud n.
// Busca r tal que 2^r >= n + 1 (propiedad de Hamming).
func calcularR(n int) int {
	r := 0
	for (1 << r) <= n {
		r++
	}
	// Retornamos el número de bits de paridad contando las potencias de 2 <= n
	count := 0
	for i := 0; (1 << i) <= n; i++ {
		count++
	}
	return count
}

// esPotenciaDe2 retorna true si n es una potencia de 2.
func esPotenciaDe2(n int) bool {
	return n > 0 && bits.OnesCount(uint(n)) == 1
}
